### PR TITLE
(Rquesting feedback, don't merge) Several new WML features and some bugfixes

### PR DESCRIPTION
**Pull request by CelticMinstrel**
First of all, don't merge this branch since it's technically incomplete (lacking changelog updates).

I'm opening this to discuss the changes herein; in particular, `[effect]apply_to=advances_to` and `[for]` have had mixed reception in #wesnoth-dev. If some features are acceptable but others aren't, I can split those off and add changelog entries for a separate pull request. (Or I suppose someone could cherry-pick them and make the changelog themselves.)

-----

I added `[for]` instead of `[foreach]` because a lot of people seem to have ideas of what `[foreach]` should be that are incompatible with what the `{FOREACH}` macro already is. The `[for]` tag as I have coded it is a "for i in range(start, stop, step)" loop, but to make it easy to substitute it for `{FOREACH}`, I have added a mode that populates start, stop, and step so that it runs over all valid indices of a WML array. Sample usages that print output to the chat area:

```
# Will print out 0, 2, 4, 6, 8, 10
[for]
  start,stop,step=0,10,2
  variable=q
  [do]
    [lua]
      code="wesnoth.message($q)"
    [/lua]
  [/do]
[/for]
# Will print out all the values in my_array, in reverse order
[for]
  array=my_array
  reverse=yes
  variable=n
  [do]
    [lua]
      code="wesnoth.message($my_array[$n].value)"
    [/lua]
  [/do]
[/for]
```

Some people have suggested omitting the `[do]` tag and just placing ActionWML straight in the `[for]` tag. This is certainly possible and may be simpler. The only reason I can think of not to do it is that having the `[do]` tag could allow a later extension using `[array]` instead of `array=` to embed the iterated array right in the for-loop, but I don't think that's actually a good use-case.

-----

For `[effect]apply_to=advances_to`, I'm open to other names; "`advances_to`" feels a bit strange. I'm also open to other names for `exclude_amla=`.

The purpose of the `[advance]` --> `[advancement]` change is so that macros defining an AMLA can also be used in `[unit][modifications]` to give a specific unit the AMLA.

### CODE REVIEW

File: data/lua/wml-tags.lua [hunk 1]
```diff
@@ -365,6 +365,16 @@ local function handle_event_commands(cfg)
...
 			arg = wesnoth.tovconfig(arg)
 		end
+		-- Interruption tags need special treatment, so they're hardcoded here
+		if cmd == "continue_if" then
+			if context_type ~= "loop" then
+				helper.wml_error("[continue] appeared outside of a loop", cmd)
+			elseif wesnoth.eval_conditional(arg) then
+				return "continue"
+			end
+		elseif cmd == "break_if" and wesnoth.eval_conditional(arg) then
```
Comment (aginor) [inline data/lua/wml-tags.lua]: Need a loop check here too, break should be invalid outside a loop context.

File: data/lua/wml-tags.lua [hunk 2]
```diff
@@ -423,12 +442,50 @@ wml_actions["while"] = function( cfg )
...
+			local action = handle_event_commands(do_child, "loop")
+			if action == "break" or action == "return" then return end
+		end
+		wesnoth.set_variable(i_var, wesnoth.get_variable(i_var) + 1)
+	end
+	end_var_scope(i_var, save_i)
+end
+
+wml_actions["repeat"] = function(cfg)
+	local times = cfg.times or 65536
```
Comment (aginor) [inline data/lua/wml-tags.lua]: Is there precedent for 65536? Anyone who tries to use this as an infinite loop will be very surprised after a long enough time.

File: data/lua/wml-tags.lua (hunk 1 above)
Comment (CelticMinstrel) [inline data/lua/wml-tags.lua]: That was intentional, actually - it means "break" can double as "return". I could disable it in `[if]` or `[switch]`, I suppose, but I probably can't disable it in a nested `[command]` block.

File: data/lua/wml-tags.lua (hunk 2 above)
Comment (CelticMinstrel) [inline data/lua/wml-tags.lua]: Yeah, the `[while]` loop has the same limit.

Comment (Vultraz) [inline data/lua/wml-tags.lua]: That number is just an arbitrary limit to prevent infinite loop hangs. It's the value of 2^16.

Comment (CelticMinstrel) [inline data/lua/wml-tags.lua]: Yes, I'm aware of that. It's still weird though. I could probably remove it in [repeat] as it's not really needed.

**Comment by CelticMinstrel** [comment]
I split `[effect]apply_to=advances_to` into a separate pull request.

**Pull request closed without merging**